### ISSUE TITLE
Libnd4j: Need loss function gradient functions

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#6517]
We have a bunch of losses here: https://github.com/deeplearning4j/deeplearning4j/tree/master/libnd4j/include/ops/declarable/generic/loss
(Plus a few others elsewhere)

These losses don't have gradient functions, which unfortunately means they aren't very useful (loss functions are only really useful for training not inference, but we need gradients for training).
This isn't urgent, but would be on the higher end of the priority spectrum.

Edit: Math in comment below.

In terms of prioritization:
High priority:
* Absolute distance
* Log loss
* Mean squared error
* Sigmoid cross entropy
* Softmax cross entropy / with logits / sparse

Medium priority:
* Cosine distance

Everything else is lower priority. (hinge, huber, mean pairwise, etc)

**Comment by AlexDBlack** [comment]
OK, so, gradients for loss functions.

We can essentially break this down to two aspects:
(a) the common reduction component - shared for multiple loss functions
(b) the per-element loss function gradient


Now, we have a couple of design choices here:
First: Do we allow calculating gradients for NONE reduction mode? All other modes have a scalar output.
Option 1 is we treat this the same as SUM mode (which would kind of make sense)
Option 2 is we only allow gradients for be calculated for the sum and mean modes

Reduction mode gradients:
L = scalar loss function value
E = element-wise loss array (the thing we sum/mean to get L depending on the mode)
p = predictions, or input to the loss function
y = labels array
w = weights array

For the weights array, I'll cover only the 'equal size' case here. For the broadcast case, we can broadcast to the approprite shape first.

SUM mode:
L = sum_i (w_i * E_i)
dL/dE_i = w_i
dL/dw_i = E_i

MEAN_BY_WEIGHT:
let W = sum_j w_j
L = 1/W * sum_i (w_i * E_i)
dL/dE_i = w_i / W
dL/dw_i = (W - w_i) / W^2

Edge case: if all w_i are 0, then L = 0 and dL/dE_i is 0

MEAN_BY_NONZERO_COUNT
L = 1/(sum_i (w_i != 0)) * sum_i (w_i * E_i)
dL/dE_i = w_i / sum_i (w_i != 0)


//////////////////////////////

Now, for the actual loss functions:
We want to calculate 3 gradients, for the 3 input arrays:
dL/dp_i
dL/dy_i
dL/dw_i

Note that (with a few exceptions, like the softmax and cosine cases) we have:
dL/dp_i = dL/dE_i * dE_i/dp_i
dL/dy_i = dL/dE_i * dE_i/dy_i
dL/dw_i = dL/dE_i * dE_i/dw_i
(note we're dropping the cross terms - i.e., assuming that dE_i/dp_j is 0, which holds for most cases, not softmax/cosine though)

Thus, we just need the dE_i/d*_i gradients for each loss function.

Mean absolute error:
E_i = abs(p_i - y_i)
dE_i/dp_i = sgn(p_i-y_i) //i.e., signum aka 1 if >=0, -1 if < 0
dE_i/dy_i = -sgn(p_i-y_i) = - dE_i/dp_i

Hinge:
Let z = 2y-1
i.e., z is labels scaled from {0,1} to {-1,1}
E_i = max(0, 1-z_i * p_i)
if(1-z_i*y_i < 0) then both gradients are 0
Otherwise
dE_i/dp_i = -z_i
dE_i/dz_i = -p_i
dE_i/dy_i = -0.5 * p_i

Huber:
E_i = 0.5 * (y_i-p_i)^2 if |y_i-p_i| <= delta
    = delta * abs(y_i-p_i) - 0.5*delta^2 otherwise

for first case:
dE_i/dy_i = y_i - p_i
dE_i/dy_i = -(y_i - p_i)

for second case:
dE_i/dy_i = delta * sgn(y_i-p_i)
dE_i/dp_i = - delta * sgn(y_i-p_i)

Log loss: (binary cross entropy)
E_i = -(y_i log(p_i + eps) + (1-y_i) log(1-p_i + eps))

dE_i/dy_i = log((1+2eps)/(p_i+eps) - 1)
dE_i/dp_i = (1-y_i)/(1+p_i+eps) - y_i/(p_i+eps)

Mean squared error:
E_i = (y_i - p_i)^2
dE_i/dy_i = 2(y_i-p_i)
dE_i/dp_i = -2(y_i-p_i)

As for the rest: I'll do another pass later.

**Comment by shyrma** [comment]
Hi, Alex

gradient functions for:
- Absolute distance
- Log loss
- Mean squared error
- Sigmoid cross entropy
- Softmax cross entropy 
- Softmax cross entropy with logits
- Sparse softmax cross entropy 
- Cosine distance
have been implemented and tested in branch shyrma_grad. This branch will be merged into master tomorrow by raver.

**Comment by AlexDBlack** [comment]
Thanks for the work here - closing this as done.
